core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -238,81 +238,83 @@ private[fs2] object Algebra {
           }
 
         case view: ViewL.View[Algebra[F, X, ?], y, Unit] =>
-          def interruptGuard(scope: CompileScope[F])(next: => F[R[X]]): F[R[X]] =
-            F.flatMap(scope.isInterrupted) {
+          def interruptGuard(guardScope: CompileScope[F])(next: => F[R[X]]): F[R[X]] =
+            F.flatMap(guardScope.isInterrupted) {
               case None => next
               case Some(Left(err)) =>
-                go(scope, view.next(Result.raiseError(err)))
+                go(guardScope, view.next(Result.raiseError(err)))
               case Some(Right(scopeId)) =>
-                go(scope, view.next(Result.interrupted(scopeId, None)))
-
+                go(guardScope, view.next(Result.interrupted(scopeId, None)))
             }
 
           view.step match {
             case output: Output[F, X] =>
-              interruptGuard(scope)(
-                F.pure(Out(output.values, scope, view.next(FreeC.Result.Pure(()))))
+              interruptGuard(goScope)(
+                F.pure(Out(output.values, goScope, view.next(FreeC.Result.unit)))
               )
 
             case u: Step[F, y, X] =>
               // if scope was specified in step, try to find it, otherwise use the current scope.
-              F.flatMap(u.scope.fold[F[Option[CompileScope[F]]]](F.pure(Some(scope))) { scopeId =>
-                scope.findStepScope(scopeId)
+              F.flatMap(u.scope.fold[F[Option[CompileScope[F]]]](F.pure(Some(goScope))) { scopeId =>
+                goScope.findStepScope(scopeId)
               }) {
                 case Some(stepScope) =>
                   F.flatMap(F.attempt(go[y](stepScope, u.stream))) {
-                    case Right(Done(scope)) =>
```

Comment:
At these point, there were three shadowed variables called `scope`: the one in this line, the one in the recursive `go` function, and the one in the entry `compileLoop`.